internal/sequencerapi/api.go

```
@@ -0,0 +1,106 @@
+package sequencerapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/internal/ethapi"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+var (
+	sendRawTxConditionalCostMeter       = metrics.NewRegisteredMeter("sequencer/sendRawTransactionConditional/cost", nil)
+	sendRawTxConditionalRequestsCounter = metrics.NewRegisteredCounter("sequencer/sendRawTransactionConditional/requests", nil)
+	sendRawTxConditionalAcceptedCounter = metrics.NewRegisteredCounter("sequencer/sendRawTransactionConditional/accepted", nil)
+)
+
+type sendRawTxCond struct {
+	b      ethapi.Backend
+	seqRPC *rpc.Client
+}
+
+func GetSendRawTxConditionalAPI(b ethapi.Backend, seqRPC *rpc.Client) rpc.API {
+	return rpc.API{
+		Namespace: "eth",
+		Service:   &sendRawTxCond{b, seqRPC},
+	}
+}
+
+func (s *sendRawTxCond) SendRawTransactionConditional(ctx context.Context, txBytes hexutil.Bytes, cond types.TransactionConditional) (common.Hash, error) {
+	sendRawTxConditionalRequestsCounter.Inc(1)
+
+	cost := cond.Cost()
+	sendRawTxConditionalCostMeter.Mark(int64(cost))
+	if cost > params.TransactionConditionalMaxCost {
+		return common.Hash{}, &rpc.JsonError{
+			Message: fmt.Sprintf("conditional cost, %d, exceeded max: %d", cost, params.TransactionConditionalMaxCost),
+			Code:    params.TransactionConditionalCostExceededMaxErrCode,
+		}
+	}
+
+	// Perform sanity validation prior to state lookups
+	if err := cond.Validate(); err != nil {
+		return common.Hash{}, &rpc.JsonError{
+			Message: fmt.Sprintf("failed conditional validation: %s", err),
+			Code:    params.TransactionConditionalRejectedErrCode,
+		}
+	}
+
+	state, header, err := s.b.StateAndHeaderByNumber(context.Background(), rpc.LatestBlockNumber)
+	if err != nil {
+		return common.Hash{}, err
```

I am still confused as to why we check the state twice. Shouldn't we do the quickest lookup to the chain tip then grab an ancestor and check both header and state against that ancestor? There is also the edge case of a short chain

---

Against HEAD we check header/state. Header is checked here since I don't see why block inclusion conditions (number/timestamp) needs to be stale to prevent MEV. State is also checked here since the transition from ancestor -> head could have invalidated the condition for account state. We could opt to not check state here and let this fail later in the pipeline at the miner.

For the ancestor, we only check the state since that's the only thing that MEV searchers would be comparing against? So the state check should also be stale


Does this help clear up the intent here? Open to not checking against HEAD and letting it fail in the miner for the edge case that ancestor -> head would have invalidated the conditional